Add XCTS system equations

nilsvu opened the pull request:
## Proposed changes

This PR adds the XCTS system equations. An upcoming PR will add the XCTS system, fluxes and sources that use these functions and that will also call into the `Poisson` and the `Elasticity` code to for the remaining operations. This is because the XCTS equations are essentially two Poisson equations and an elasticity equation on a curved geometry, coupled through nonlinear sources.

Dependencies:

- [x] #2876 

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu commented:
@markscheel @haraldp271 @geoffrey4444 @nilsdeppe Here, at last, are the XCTS equations. Would two of you be willing to review the code? The PR is, of course, quite big and almost exclusively tensor-math 🤯😄

nilsdeppe commented:
This seems like it should be 2-3 PRs... There are several "short" commits (200-400 lines) and then the giant XCTS one. Does it really all need to be one monster PR?

nilsvu commented:
Sure, I can factor the Poisson and Elasticity commits into separate PRs. I kept them here because they provide terms for the XCTS equations and it might actually be better to review them in the same context. It's a lot of math either way ^^ I'll do what reviewers prefer

nilsdeppe commented:
So I at least would be able to review the non-XCTS part of this PR fairly "casually" and not feel like I'm committing so something huge (so on the time scale of a few days, maybe faster)

nilsvu commented:
Ok here's the first 3 commits as a separate PR: #2674

nilsvu commented:
Also factored out the fourth commit here: #2675

haraldp271 commented on src/Elliptic/Systems/Elasticity/Equations.cpp:
```diff
@@ -55,6 +83,21 @@ void auxiliary_fluxes(
+  for (size_t d = 0; d < Dim; d++) {
+    flux_for_strain->get(d, d, d) += 0.5 * co_displacement.get(d);
+    for (size_t e = 0; e < Dim; e++) {
+      flux_for_strain->get(d, e, d) += 0.5 * co_displacement.get(e);
```
These are a lot of duplicated multiplications by 0.5.  Can line 91 be written as 

`const auto co_displacement_over2 = 0.5*raise_or_lower_index(displacement, metric);
`

nilsvu replied:
This code was merged in #2674. I think it's better to limit this PR to the XCTS equations now. I'm happy to go over the elasticity code in a separate PR to fix these kinds of things though.

haraldp271 commented on src/Elliptic/Systems/Poisson/Equations.cpp:
```diff
@@ -27,12 +28,17 @@ template <size_t Dim>
+    const gsl::not_null<Scalar<DataVector>*> source_for_field,
+    const tnsr::i<DataVector, Dim>& christoffel_contracted,
+    const tnsr::I<DataVector, Dim>& flux_for_field) noexcept {
+  get(*source_for_field) -=
```
I find the differing syntax for get surprising.  Why not `source_for_field->get()` ?

Probably a design decision already discussed elsewhere, if so, no long explanation necessary.

nilsvu replied:
Our tensors have two kinds of `get` functions: (1) Member functions like `tensor.get(i, j)` for runtime-access to the components and (2) free functions like `get<0, 1>(tensor)` for when the components are known at compile-time. Since a `Scalar` only has a single component we can always use the compile-time version `get(scalar)`.

haraldp271 commented on src/Elliptic/Systems/Poisson/Equations.hpp:
```diff
@@ -29,16 +36,28 @@ void euclidean_fluxes(gsl::not_null<tnsr::I<DataVector, Dim>*> flux_for_field,
 
 /*!
- * \brief Compute the fluxes \f$F^i=\sqrt{\gamma}\gamma^{ij}\partial_j u(x)\f$
+ * \brief Compute the fluxes \f$F^i=\gamma^{ij}\partial_j u(x)\f$
```
It seems that `Poisson` refers to a **scalar** Poisson equation only.  That's fine, but please make sure this is mentioned near the start of the primary Poisson documentation.

nilsvu replied:
This code was also merged in #2674

haraldp271 commented on src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp:
```diff
@@ -0,0 +1,247 @@
+ * the symmetric shift strain \f$B_{ij}=\bar{D}_{(i}\beta_{j)}\f$ and the
+ * gradient of the lapse times the conformal factor
+ * \f$w_i=\partial_i\left(\alpha\psi\right)\f$. When we then cast the equations
+ * in first-order flux-form
```
```suggestion
 * \f$w_i=\partial_i\left(\alpha\psi\right)\f$.  Please note that $B_{ij}$ includes its  
 * trace-part, i.e. it differs from the longitudinal operator $(L\beta)_{ij}$ from 
 * the NR literature. 
 *
 * When we cast the equations in first-order flux form
```

haraldp271 commented on src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp:
```diff
@@ -0,0 +1,247 @@
+ *
+ * Note that the symbol \f$\beta\f$ in the equations above means
+ * \f$\beta_\mathrm{excess}\f$. The full shift is \f$\beta_\mathrm{excess} +
+ * \beta_\mathrm{background}\f$. Also note that the background shift is
```
Give a pointer to what beta_background might be.  Guessing, perhaps something like `... where \f$beta_\mathrm{background}\f$ is generally \f$\Omega \ times \vec{r} + dota \vec{r} \f$`

nilsvu replied:
This example is already in the documentation of the `ShiftExcess` tag, so I added a reference to that documentation.

haraldp271 commented on src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp:
```diff
@@ -0,0 +1,247 @@
+ *
+ * \par Conformal background geometry:
+ * The equations simplify significantly if the conformal metric is flat
+ * ("conformal flatness"). In this case you can specify
```
```suggestion
 * The equations simplify significantly if the conformal metric is flat and Cartesian 
  * coordinates are used ("conformal flatness"). In this case you can specify
```

haraldp271 commented on src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp:
```diff
@@ -0,0 +1,247 @@
+                      Tags::ConformalChristoffelSecondKind<DataVector, 3,
+                                                           Frame::Inertial>>,
+                  tmpl::list<>>>,
+          tmpl::list<>>>>;
```
Well, I wonder whether all of this is overdoing templates.  But I trust you and other spectre developers have discussed the level of templating already more than enough.

haraldp271 commented on src/Elliptic/Systems/Xcts/Geometry.hpp:
```diff
@@ -0,0 +1,9 @@
+namespace Xcts {
+/// \brief Euclidean or non-Euclidean background geometry
+enum class Geometry { Euclidean, NonEuclidean };
+}  // namespace Poisson
```
```suggestion
}  // namespace Xcts
```

haraldp271 commented on src/Elliptic/Systems/Xcts/Geometry.hpp:
```diff
@@ -0,0 +1,9 @@
+
+namespace Xcts {
+/// \brief Euclidean or non-Euclidean background geometry
+enum class Geometry { Euclidean, NonEuclidean };
```
Is it smart to have a separate Xcts::Geometry that duplcaites Poisson::Geometry?   After all, the two must be used and modified always in lockstep.

nilsvu replied:
Not sure, the actual code duplication here is minimal since the enums mostly exist for documentation and to write `FlatCartesian`/`Curved` instead of booleans. I think it might be better to keep the enum and its documentation specific to the system instead of trying to generalize it too much. Either could work though 🤷‍♂️

haraldp271 commented on src/Elliptic/Systems/Xcts/Tags.hpp:
```diff
@@ -12,7 +12,63 @@
+ * We solve the XCTS equations for the conformal factor \f$\psi\f$, the product
+ * of lapse times conformal factor \f$\alpha\psi\f$ and the shift vector
+ * \f$\beta\f$. The remaining quantities in the equations, i.e. the conformal
+ * metric \f$\bar{\gamma}_{ij}\f$, the extrinsic curvature \f$K\f$, their
```
```suggestion
 * metric \f$\bar{\gamma}_{ij}\f$, the trace of the extrinsic curvature \f$K\f$, their
```

haraldp271 commented on src/Elliptic/Systems/Xcts/Tags.hpp:
```diff
@@ -12,7 +12,63 @@
+ * \cite BaumgarteShapiro, in particular Box 3.3 which is largely mirrored here.
+ * We solve the XCTS equations for the conformal factor \f$\psi\f$, the product
+ * of lapse times conformal factor \f$\alpha\psi\f$ and the shift vector
+ * \f$\beta\f$. The remaining quantities in the equations, i.e. the conformal
```
```suggestion
 * \f$\beta^i\f$. The remaining quantities in the equations, i.e. the conformal
```

haraldp271 commented on src/Elliptic/Systems/Xcts/Tags.hpp:
```diff
@@ -12,7 +12,63 @@
+ * of lapse times conformal factor \f$\alpha\psi\f$ and the shift vector
+ * \f$\beta\f$. The remaining quantities in the equations, i.e. the conformal
+ * metric \f$\bar{\gamma}_{ij}\f$, the extrinsic curvature \f$K\f$, their
+ * respective time derivatives \f$\bar{u}_{ij}\f$ and \f$\partial_t K\f$, the
```
elsewhere I saw a comment that u_{ij} and beta_background are combined into one field.  Mention here (or somewhat further down)

haraldp271 commented on src/Elliptic/Systems/Xcts/Tags.hpp:
```diff
@@ -12,7 +12,63 @@
+ * scenario at hand. Of particular importance is the conformal metric, which
+ * serves as the background geometry for these equations by defining the
+ * covariant derivative \f$\bar{D}\f$, the Ricci scalar \f$\bar{R}\f$ and the
+ * longitudinal operator
```
```suggestion
 * scenario at hand. Of particular importance is the conformal metric, which
 * defines the background geometry, the covariant derivative \f$\bar{D}\f$, 
  * the Ricci scalar \f$\bar{R}\f$ and the longitudinal operator
```

haraldp271 commented on src/Elliptic/Systems/Xcts/Tags.hpp:
```diff
@@ -12,7 +12,63 @@
+ * \f}
+ *
+ * Note that the XCTS equations are essentially two Poisson equations and one
+ * Elasticity equation with nonlinear sources on a non-Euclidean geometry.
```
One difference for the shift equation is that in the GR-shift equation, the trace of the gradient is removed (i.e. longitudinal operator vs. strain).   State explicitly whether this matters, and if so how.

nilsvu replied:
The longitudinal operator is actually just analogous to a particular elastic constitutive relation, I added a few notes to the docs.

haraldp271 commented on src/Elliptic/Systems/Xcts/Tags.hpp:
```diff
@@ -121,5 +186,90 @@ struct ShiftStrain : db::SimpleTag {
+};
+
+/*!
+ * \brief The shift vector contracted with the extrinsic curvature trace
+ * gradient: \f$\beta^i\partial_i K\f$
```
```suggestion
 * \brief The shift vector contracted with the gradient of the trace 
 * of the extrinsic curvature: \f$\beta^i\partial_i K\f$
```

haraldp271 commented on src/PointwiseFunctions/GeneralRelativity/Tags.hpp:
```diff
@@ -105,6 +105,12 @@ template <size_t Dim, typename Frame, typename DataType>
+/// Contraction of the first two indices of the spatial Christoffel symbols:
+/// \f$\Gamma^i_{ij}\f$. Useful for covariant divergences.
+template <size_t Dim, typename Frame, typename DataType>
+struct SpatialChristoffelSecondKindContracted : db::SimpleTag {
```
@markscheel @kidder Is this the most sensible name for Gamma_i, or is there already a different standardized name in the KST papers, SpEC, or elsewhere?

haraldp271 replied:
Since the Christoffel symbols of the second kind are the standard ones nearly always used, I would suggest to just call them 'Christoffel' for brevity.   In the fewer cases where Gamma_{ijk} is needed, those can be called 'ChristoffelFirstKind'.

haraldp271 reviewed (commented):
Some early feedback, mostly on comments.  
- this PR still has a lot of code from Poisson and Elasticity, which I believe is part of a separate PR, so I didn't comment on it here; will this Xcts PR be reposed in smaller form once Poisson i s merged?
- I also did NOT check the algebra or the equations.  Happy to do so next time, let me know.
- I also did NOT check the algebra in the tests -- are people worried about the same algebraic error being present in both the tests and the real code (after all, this is based  on the same calc of Nils...)

moxcodes commented:
All of the PR's mentioned here appear to have been merged -- @nilsleiffischer is this ready to be rebased and reviewed as non-dependent?

nilsvu commented:
It is, and I'll get this rebased ASAP

nilsvu commented:
Factored out a few more PRs and updated the list of dependencies in the PR description :)

nilsvu commented:
@haraldp271 I have rebased this PR on latest develop and added a fixup commit to address your comments. I think this is ready for review now. Please take a look when you get the chance!

<s>In your review you can ignore the first two commits, since they have their own PRs (#2875 and #2876) and will disappear once those are merged.</s> Those two PRs are merged now, but I factored a bit more code out into to make this PR smaller. #2915 just adds the tags and some documentation.

nilsvu committed:
Add enum for XCTS conformal background geometry

nilsvu committed:
Add XCTS equations

nilsvu commented:
@haraldp271 I stripped this PR down _a lot_ so it only adds the actual math. The next PR will add the code that ties everything together. Please continue your review here when you get the chance!

haraldp271 commented on src/Elliptic/Systems/Xcts/Equations.hpp:
```diff
@@ -0,0 +1,252 @@
+// expressions once those work
+void fully_contract(gsl::not_null<Scalar<DataVector>*> result,
+                    const tnsr::II<DataVector, 3>& tensor1,
+                    const tnsr::II<DataVector, 3>& tensor2) noexcept;
```
The index-placement of the tensors is surprising.  Perhaps add a _flat to the function name?

haraldp271 commented on src/Elliptic/Systems/Xcts/Equations.hpp:
```diff
@@ -0,0 +1,252 @@
+ *
+ * Adds \f$\frac{1}{12}\psi^5 K^2 - 2\pi\psi^5\rho\f$. Additional sources can be
+ * added with `add_distortion_hamiltonian_sources` and
+ * `add_curved_hamiltonian_or_lapse_sources`.
```
The matter term is problematic, because the unscaled rho.  See https://arxiv.org/abs/gr-qc/0610120, discussion after Eq. (2), and in fact the entire paper, which is about those problems.

Recommend you open an issue to improve on the rho scaling in the future.  E.g. code as psi^{5-n} \tilde{\rho},  where \tilde{\rho}=\psi^n \rho is now the freely specifiable data, and let the user specify the value for n.  (n=8 was suggested by York, as it preseves a certain energy condition;  n=6 might also be useful, as it preserves the total energy density \sqrt{g}rho  =  \sqrt{tilde{g}} \tilde{rho} )

nilsvu replied:
I just went ahead and added the conformal scaling of the matter terms. I pushed the change as a fixup commit because it's not entirely trivial.

haraldp271 commented on src/Elliptic/Systems/Xcts/Equations.hpp:
```diff
@@ -0,0 +1,252 @@
+ *
+ * Note that this linearization is only w.r.t. \f$\psi\f$.
+ */
+void add_linearized_distortion_hamiltonian_sources(
```
Perhaps this is handled elsewhere (and I'll discover as I go on): 
A_{ij} contains the shift, so should there also be  shift_correction?

nilsvu replied:
You probably discovered later that the shift correction is handled by `add_*_momentum_sources`. This is so the functions can also be used when not solving for the shift.

haraldp271 commented on src/Elliptic/Systems/Xcts/Equations.hpp:
```diff
@@ -0,0 +1,252 @@
+ * divergence, but only the partial-derivatives part of it. The curved
+ * contribution to this term can be added together with the curved contribution
+ * to the flux divergence of the dynamic shift variable with the
+ * `Elasticity::add_curved_sources` function.
```
Elasticity?

nilsvu replied:
Yep, the elasticity function can be used here as-is. It just adds the Christoffel-terms for the divergence of a symmetric rank-two tensor field (the elastic stress / longitudinal shift)

haraldp271 commented on src/Elliptic/Systems/Xcts/Equations.cpp:
```diff
@@ -0,0 +1,619 @@
+      for (size_t k = 0; k < 3; ++k) {
+        for (size_t l = 0; l < 3; ++l) {
+          get(*result) += metric.get(i, k) * metric.get(j, l) *
+                          tensor1.get(i, j) * tensor2.get(k, l);
```
four nested loops, with three multiplies in the innermost loop gives 3^4*3 = 243 multiplies.

If one had first raised an index,  $T^i_j = g^{ik} T_{kj}$  (9*3 = 27 multiplies), and then contracted   $T^i_j\,T^j_i$   (9 multiplies), then only 36 multiplies in total would be needed, but also a temporary. 

At some point it would be interesting to time where the cross-over is between temporary vs. more floating point operations.

nilsvu replied:
I actually have some temporary memory to spare at the points where this function is used, so I refactored it to perform 63 multiplications instead of the 243 (63 because the two tensors may not be the same) without allocating more memory.

haraldp271 commented on src/Elliptic/Systems/Xcts/Equations.cpp:
```diff
@@ -0,0 +1,619 @@
+      ((5. / 12.) * square(get(extrinsic_curvature_trace)) -
+       10. * M_PI * get(energy_density)) *
+      pow<4>(get(conformal_factor)) * get(conformal_factor_correction);
+}
```
SpEC pre-computes terms in the linearization.  For this specific example, SpEC would compute and store   A = (5/12 K^2 + 10*pi*rho)*psi^4.    This is done *once* per non-linear Newton-Raphson iteration.

The linearization is then evaluated as   += A*conformal_factor_correction.    This (cheap) evaluation is then evaluated at every linear iteration.

Would be interesting at which point this becomes advantageous.  Clearly, this will depend on the number of linear iterations typically taken, and the relative cost of evaluating the linearization vs. applying the preconditioner.

nilsvu replied:
Opened https://github.com/sxs-collaboration/spectre/issues/3071 for this. Pre-computing these quantities could potentially speed up current solves because they explicitly build the subdomain operator by evaluating the linearization many times, but the performance gain might decrease with better preconditioners.

haraldp271 commented on src/Elliptic/Systems/Xcts/Equations.hpp:
```diff
@@ -0,0 +1,252 @@
+ * \brief Add the "distortion" source term to the Hamiltonian constraint.
+ *
+ * Adds \f$-\frac{1}{8}\psi^{-7}\bar{A}^{ij}\bar{A}_{ij}\f$.
+ */
```
Please add formula defining longitudinal_shift_minus_dt_conformal_metric_over_lapse_square

nilsvu replied:
Added references to the tag that defines this quantity (and `longitudinal_shift_minus_dt_conformal_metric_square` as well).

haraldp271 commented on src/Elliptic/Systems/Xcts/Equations.cpp:
```diff
@@ -0,0 +1,619 @@
+}
+
+template <Geometry ConformalGeometry>
+void add_momentum_sources_impl(
```
The formulas in this function (and some others) are complicated enough that references to concrete equation numbers of your favourite paper/book would be useful.  Please add something like 
"This function implements Eqs. (xx) of Baumgarte+Shapiro "Numerical Relativity"  at the relevant places.

haraldp271 reviewed (dismissed):
My comments are trivial enough that I trust Nils to implement them before merging, w/o me seeing this again.

nilsvu commented:
Thanks for reviewing @haraldp271! I pushed a few fixup commits since they were not entirely non-trivial. Please take a look at the changes when you get the chance.

haraldp271 commented on src/Elliptic/Systems/Xcts/Geometry.hpp:
```diff
@@ -0,0 +1,18 @@
+
+namespace Xcts {
+/// \brief Types of conformal geometries for the XCTS equations
+enum class Geometry {
```
I don't understand why we need separate, identical tags in Poisson and Xcts.  Wouldn't code re-use point toward just having one Geometry tag?   And also, are separate Poisson::Geometry and Xcts::Geometry tags going to cause inconsistencies later on?  E.g. what if Geometry-templated XCTS code wants to re-use Poisson-code?

nilsvu replied:
The enums could be merged, though I didn't do that so far because:

1. I don't think the code re-use is super important here since it's just an enum to get human-readable labels and there's no logic involved. Generalizing and re-using code is nice, but taken too far it can make the code more difficult to follow.
2. I'm not sure where to place it in the codebase since `elliptic::Geometry` doesn't seem right; it could be more general than the elliptic code, but such generality is probably unnecessary.
3. I don't think inconsistencies with Poisson code will be an issue. The Poisson (and elasticity) code is called internally to re-use implementations, but none of that is exposed to the public interface. Also the Poisson and elasticity functions are called directly (e.g. `Poisson::curved_fluxes(...)`), not using the `Poisson::Geometry` enum. The enum is there just to have nicer labels for the template parameters.

haraldp271 replied:
I am fine if you're fine.

nilsvu commented:
@nilsdeppe would you be willing to give this a second review?

nilsdeppe commented:
Sure!

nilsdeppe commented on src/Elliptic/Systems/Xcts/Equations.hpp:
```diff
@@ -0,0 +1,283 @@
+ * "distortion" source term to the Hamiltonian constraint and lapse equation.
+ *
+ * Adds \f$\left((\bar{L}\beta)^{ij} -
+ * \bar{u}^{ij}\right)\left(\frac{w_j}{\alpha\psi} - 7 \frac{v_j}{\psi}\right)
```
could you please document what `w_j` and `v_j` are?

nilsvu replied:
They are defined in the docs for the XCTS system which isn't in this PR yet, so I replaced them with `\partial_j (\alpha\psi)` and `\partial_j \psi`.

nilsdeppe commented on src/Elliptic/Systems/Xcts/Equations.cpp:
```diff
@@ -0,0 +1,779 @@
+    const Scalar<DataVector>& conformal_factor) noexcept {
+  // See BaumgarteShapiro, Eq. (3.110) and (3.112)
+  get(*hamiltonian_constraint) -=
+      0.03125 * pow<5>(get(conformal_factor)) *
```
[optional] I have no objection to having numbers like `0.03125` here, but it would be nice to have a comment of what fraction it is :) If you decide to add a comment for this here, could you also do so in the below functions?

nilsvu replied:
done 👍

nilsdeppe reviewed (dismissed):
LGTM, couple small documentation requests. Please squash if you agree with the suggestions :)

nilsvu commented:
Squashed in the added comments and rebased. Thanks for reviewing @haraldp271 and @nilsdeppe!

nilsvu commented:
@haraldp271 Could you sign off on this before we merge it? :)

haraldp271 reviewed (approved):
my comments are taken care of.  So happy to have this merged.
